File: binder_pool_client_b/src/main/java/com/demo/binder_pool_client_b/MainActivity.java
package com.demo.binder_pool_client_b;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;
import android.view.View;

import com.demo.binder_pool_server.IBinderPool;
import com.demo.binder_pool_server.IMinusOperation;

public class MainActivity extends Activity {
    private IMinusOperation minusOperation;
    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            IBinderPool binderPool = IBinderPool.Stub.asInterface(iBinder);
            try {
                IBinder binder = binderPool.getBinder(200);
                minusOperation = IMinusOperation.Stub.asInterface(binder);
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            minusOperation = null;
        }
    };

    public void onMinusBtnClick(View view) {
        try {
            if (minusOperation != null) {
                int result = minusOperation.minus(5, 2);
                Log.d("gxd", "5 - 2 = " + result);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    private void bindService() {
        Intent intent = new Intent();
        intent.setClassName("com.demo.binder_pool_server", "com.demo.binder_pool_server.BinderPoolService");
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        bindService();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (minusOperation != null) {
            unbindService(serviceConnection);
        }
    }
}

File: binder_pool_client_minus/src/main/java/com/gxd/demo/aidl/client/minus/MainActivity.kt
package com.gxd.demo.aidl.client.minus

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.Process
import android.util.Log
import android.widget.Button
import com.gxd.demo.aidl.IBinderPool
import com.gxd.demo.aidl.IMinusOperation

class MainActivity : Activity() {
    private var minusOperation: IMinusOperation? = null
    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName?, iBinder: IBinder?) {
            val binderPool = IBinderPool.Stub.asInterface(iBinder)
            val binder = binderPool.getBinder(200)
            minusOperation = IMinusOperation.Stub.asInterface(binder)
        }

        override fun onServiceDisconnected(componentName: ComponentName?) {
            minusOperation = null
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val intent = Intent().apply {
            val packageName = "com.gxd.demo.aidl.pool.server"
            setClassName(packageName, "$packageName.BinderPoolService")
            setPackage(packageName)
        }
        val bindResult = bindService(intent, serviceConnection, BIND_AUTO_CREATE)
        "bindResult = $bindResult".log()

        val button = Button(this)
        with(button) {
            button.text = "Minus"
            button.setOnClickListener {
                val result = minusOperation?.minus(5, 2)
                "5 - 2 = $result".log()
            }
        }
        setContentView(button)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (minusOperation != null) unbindService(serviceConnection)
    }

    private val logPrefix: String
        get() = "Client...pid = ${Process.myPid()}, thread = ${Thread.currentThread().name}-${Thread.currentThread().id}"

    private fun String.log() {
        Log.d("ggg", "$logPrefix, $this")
    }
}